### PR TITLE
VOL-126: Proximity searching for volunteer opportunities

### PR DESCRIPTION
**Pull request by GinkgoFJG**

### CODE REVIEW

File: CRM/Volunteer/BAO/Project.php [hunk 1]
```diff
@@ -237,59 +237,166 @@ static function isActive($entityId, $entityTable) {
    * Get a list of Projects matching the params.
    *
    * This function is invoked from within the web form layer and also from the
-   * API layer. Params keys are either column names of civicrm_volunteer_project, or
-   * 'project_contacts.' @see CRM_Volunteer_BAO_Project::create() for details on
-   * this parameter.
+   * API layer. Special params include:
+   * <ol>
+   *   <li>project_contacts (@see CRM_Volunteer_BAO_Project::create() and
+   *     CRM_Volunteer_BAO_Project::buildContactWhere)</li>
```
Comment (ginkgomzd) [inline CRM/Volunteer/BAO/Project.php]: I believe this function is called buildContactJoin

File: CRM/Volunteer/BAO/Project.php [hunk 2]
```diff
@@ -237,59 +237,166 @@ static function isActive($entityId, $entityTable) {
...
+    $result = FALSE;
+    $onClauses = array();
+
+    $relTypes = CRM_Core_OptionGroup::values(
+      CRM_Volunteer_BAO_ProjectContact::RELATIONSHIP_OPTION_GROUP,
+      TRUE, FALSE, FALSE, NULL, 'name');
 
-    while ($project->fetch()) {
-      $result[(int) $project->id] = clone $project;
+    foreach ($projectContacts as $relType => $contactIds) {
```
Comment (ginkgomzd) [inline CRM/Volunteer/BAO/Project.php]: foreach .... {
  $subquery[] = "SELECT $contact_id as `cid`, $relationship_type_id as `rel`":
}

$strSubquery = "(" . join("\nUNION\n", $subquery) . ") vpc_ids";

$join = $strSubquery . " INNER JOIN vpc on vpc_ids.cid = vp.contact_id AND vpc.relationship_type_id = vpc_ids.rel ";


... don't know for sure which approach would be more performant... depends on how smart the compiler is at unraveling your join condition.

Is there a limit on how many contacts can be related?

File: CRM/Volunteer/BAO/Project.php [hunk 3]
```diff
@@ -237,59 +237,166 @@ static function isActive($entityId, $entityTable) {
...
+    if (!CRM_Utils_Rule::numeric($radius)) {
+      throw new Exception(ts('Radius should exist and be numeric'));
+    }
+
+    if (!CRM_Utils_Rule::numeric($lat) || !CRM_Utils_Rule::numeric($lon)) {
+      if (empty($country)) {
+        throw new Exception(ts('Either Country or both Latitude and Longitude are required'));
+      }
+
+      // TODO: I think CRM_Utils_Geocode_*::format should be responsible for this
```
Comment (ginkgomzd) [inline CRM/Volunteer/BAO/Project.php]: Is there an issue in jira or this?

File: CRM/Volunteer/BAO/Project.php [hunk 4]
```diff
@@ -237,59 +237,166 @@ static function isActive($entityId, $entityTable) {
...
+      if (CRM_Utils_Type::validate($country, 'Positive', FALSE)) {
+        $country = civicrm_api3('Country', 'getvalue', array(
+          'id' => $country,
+          'return' => 'name',
+        ));
+      }
+
+      // TODO: support other geocoders
```
Comment (ginkgomzd) [inline CRM/Volunteer/BAO/Project.php]: Is there an issue in jira for this?

File: CRM/Volunteer/BAO/Project.php (hunk 1 above)
Comment (GinkgoFJG) [inline CRM/Volunteer/BAO/Project.php]: Thanks. It'd been renamed in refactoring.

File: CRM/Volunteer/BAO/Project.php (hunk 2 above)
Comment (GinkgoFJG) [inline CRM/Volunteer/BAO/Project.php]: There is not a limit, but the relationship types are Project Manager, Project Owner, and Project Beneficiary -- each can take multiple contacts, but I don't imagine that happening too often, and not with large numbers of contacts when it does.

File: CRM/Volunteer/BAO/Project.php (hunk 3 above)
Comment (GinkgoFJG) [inline CRM/Volunteer/BAO/Project.php]: Per your request, I created this: [CRM-17245](https://issues.civicrm.org/jira/browse/CRM-17245).

File: CRM/Volunteer/BAO/Project.php (hunk 4 above)
Comment (GinkgoFJG) [inline CRM/Volunteer/BAO/Project.php]: Yes, it's VOL-136.

**Pull request merged**